src/main.cpp

```diff
@@ -465,16 +478,16 @@ void UpdateBlockAvailability(NodeId nodeid, const uint256 &hash) {
 }
 
 void MaybeSetPeerAsAnnouncingHeaderAndIDs(const CNodeState* nodestate, CNode* pfrom, CConnman& connman) {
-    if (pfrom->GetLocalServices() & NODE_WITNESS) {
-        // Don't ever request compact blocks when segwit is enabled.
+    if (!nodestate->fSupportsDesiredCmpctVersion) {
```

Comment:
A node is never prevented from chosing to send us a message, but later in this function is the only place we send a sendcmpct with announce set to true, so no matter how you call this it should be safe (note that we set fSupportsDesiredCmpctVersion based on NODE_WITNESS, not based on current activation status).

As for the duplicate check...whatever, up to the PR owner, I suppose :p.